File: 1-two-sum/solution.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # Create a dictionary to store numbers and their indices
        num_map = {} 

        # Iterate through the array with both index and value
        for i, num in enumerate(nums):
            # Calculate the complement needed to reach the target
            complement = target - num

            # Check if the complement exists in the map
            if complement in num_map:
                # If found, return the index of the complement and the current index
                return [num_map[complement], i]
            
            # If not found, add the current number and its index to the map
            num_map[num] = i

        # This part should ideally not be reached if a solution is guaranteed
        return [] 

File: 1-two-sum/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // Hash map approach - O(n) time, O(n) space
    const numMap = new Map();
    
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        
        // Check if complement exists in our map
        if (numMap.has(complement)) {
            return [numMap.get(complement), i];
        }
        
        // Store current number and its index
        numMap.set(nums[i], i);
    }
    
    return []; // This shouldn't happen based on problem constraints
};

// Alternative brute force solution (less efficient)
var twoSumBruteForce = function(nums, target) {
    // O(n²) time, O(1) space
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
    return [];
};

// Test cases
console.log("Test Case 1:", twoSum([2,7,11,15], 9)); // Expected: [0,1]
console.log("Test Case 2:", twoSum([3,2,4], 6));     // Expected: [1,2]  
console.log("Test Case 3:", twoSum([3,3], 6));       // Expected: [0,1]